api/src/main/java/org/apache/iceberg/expressions/ResidualEvaluator.java

```diff
@@ -211,31 +210,48 @@ public Expression or(Expression leftResult, Expression rightResult) {
         return pred; // not associated inclusive a partition field, can't be evaluated
       }
 
-      List<UnboundPredicate<?>> strictProjections = Lists.transform(parts,
-          part -> ((Transform<T, ?>) part.transform()).projectStrict(part.name(), pred));
+      for (PartitionField part : parts) {
```

Comment:
I like how you restructured this to simply return `alwaysTrue` or `alwaysFalse` if any projection can determine the result. That's a lot simpler than before.